### PR TITLE
[NVIDIA] Support the variable sequence lengths in `jax.nn.dot_product_attention`

### PR DESCRIPTION
**Pull request by kaixih**
This PR enhances the `jax.nn.dot_product_attention` API to support the variable sequence lengths.

It introduces two additional arguments of `q_sequence_lengths` and `kv_sequence_lengths` and the backend could avoid doing computations for the padding values.

The PR is on top of [this PR](https://github.com/google/jax/pull/22477).

### CODE REVIEW

**Comment by kaixih** [comment]
@sbodenstein can you help review?

**Comment by kaixih** [comment]
Also, cc @superbobry

File: jax/_src/nn/functions.py [hunk 1]
```diff
@@ -903,7 +935,11 @@ def dot_product_attention(
       logits to mask out the non-causal parts of the attention matrix, but other
       implementations like `cudnn` will avoid computing the non-causal regions,
       providing speedups.
-    implementation: A string to control which implementation backend to use.
+    q_sequence_lengths: `int32` array of sequence lengths for query; shape
+      :code:`(B)`
+    kv_sequence_lengths: `int32` array of sequence lengths for key and value;
+      shape :code:`(B)`
+    implementaion: A string to control which implementation backend to use.
```
[the remark this answers is not in the document]
Comment (kaixih) [reply jax/_src/nn/functions.py]: Done.

File: jax/_src/nn/functions.py [hunk 2]
```diff
@@ -903,7 +935,11 @@ def dot_product_attention(
       logits to mask out the non-causal parts of the attention matrix, but other
       implementations like `cudnn` will avoid computing the non-causal regions,
       providing speedups.
-    implementation: A string to control which implementation backend to use.
+    q_sequence_lengths: `int32` array of sequence lengths for query; shape
```
[the remark this answers is not in the document]
Comment (kaixih) [reply jax/_src/nn/functions.py]: Done.

**Comment by sbodenstein** [comment]
LGTM

**Comment by kaixih** [comment]
Just squashed the commits. @sbodenstein @superbobry Can you help merge? Thx.

**Review by dfm — APPROVED** [review]

**Comment by dfm** [comment]
@kaixih — @superbobry is OOO, but I can help get this landed. Can you start by taking a look at the linting and typing failures? It can be useful to run `pre-commit` locally ([docs](https://jax.readthedocs.io/en/latest/contributing.html#linting-and-type-checking)) to make sure that's all green first. Then, don't forget to squash commits. Thanks!!

**Comment by kaixih** [comment]
@dfm Thanks for taking this on. I've used pre-commit to check the lint format. PTAL.

**Review by dfm — APPROVED** [review]
